libraries/lib-uuid/Uuid.cpp

```diff
@@ -0,0 +1,216 @@
+/*!********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file Uuid.cpp
+ @brief Define a class to generate and parse UUIDs
+
+ Dmitry Vedenko
+ **********************************************************************/
+
+/*!********************************************************************
+
+ @class Uuid
+ @brief Platform independent class for generating and parsing UUIDs.
+
+ **********************************************************************/
+
+#include "Uuid.h"
+
+#include <algorithm>
+#include <cstring>
+
+#if defined(USE_UUID_CREATE)
+#   include <rpc.h>
+#elif defined(USE_CFUUID)
+#   include <CoreFoundation/CFUUID.h>
+#elif defined(USE_LIBUUID)
+#   include <uuid/uuid.h>
+#endif
+
+#include "HexHelpers.h"
+
+namespace audacity
+{
+
+namespace
+{
+bool readByte(Uuid::Bytes::iterator& outputIt,
+              std::string::const_iterator& inputIt,
+              const std::string::const_iterator& inputEnd)
+{
+   if (inputIt == inputEnd)
+      return false;
+
+   const char c1 = *inputIt++;
+
+   if (!std::isxdigit(c1))
+      return false;
+
+   if (inputIt == inputEnd)
+      return false;
+
+   const char c2 = *inputIt++;
+
+   if (!std::isxdigit(c2))
+      return false;
+
+   *outputIt = uint8_t((HexCharToNum(c1) << 4) | HexCharToNum(c2));
+
+   ++outputIt;
+
+   return true;
+}
+} // namespace
+
+Uuid::Uuid()
+    : Uuid(Bytes {})
+{
+}
+
+Uuid::Uuid(const Bytes& data) noexcept
+    : mData(data)
+{
+}
+
+Uuid Uuid::Generate()
+{
+#if defined(USE_UUID_CREATE)
+   UUID winUid;
+
+   if (RPC_S_OK != ::UuidCreate(&winUid))
+      return {};
+
+   Uuid uuid;
+
+   std::memcpy(uuid.mData.data(), &winUid, sizeof(winUid));
+
+   return uuid;
+#elif defined(USE_CFUUID)
+   CFUUIDRef newId = CFUUIDCreate(NULL);
+   CFUUIDBytes bytes = CFUUIDGetUUIDBytes(newId);
+   CFRelease(newId);
+
+   Uuid uuid;
+
+   std::memcpy(uuid.mData.data(), &bytes, sizeof(bytes));
+
+   return uuid;
+#elif defined(USE_LIBUUID)
+   uuid_t newId;
+
+   uuid_generate(newId);
+
+   Uuid uuid;
+
+   std::memcpy(uuid.mData.data(), newId, sizeof(newId));
+
+   return uuid;
+#else
+#   error "UUID generator is not defined"
+#endif
+}
+
+Uuid Uuid::FromString(const std::string& str)
+{
+   const size_t length = str.length();
+
+   if (length == 0)
+      return {};
+
+   const bool hasBraces = str[0] == '{';
+
+   if (hasBraces && (length != 38 || str.back() != '}'))
+      return {};
+   else if (!hasBraces && length != 36)
+      return {};
```

Comment:
You require that str is exactly a Uuid or a brace-enclosed one, else return nil.

You aren't writing a function that just allows a prefix of a string to be a Uuid.

Comment that.  Though maybe I am supposed to guess that because the function isn't also returning a std::string::const_iterator.

Reply:
When I ask for comments, I mean in the header.  Because it's about the contract, not just the implementation.